### ISSUE TITLE
Constrained Policy Optimization  The Berkeley Artificial Intelligence Research Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1210]
Constrained Policy Optimization – The Berkeley Artificial Intelligence Research Blog<br>
<br>
https://ift.tt/2sKqzjW<br>
<br>
Daniel Seita<br>
<br>

<div>
<article>
  <p>(Based on joint work with David Held, Aviv Tamar, and Pieter Abbeel.)</p>

<p>Deep reinforcement learning (RL) has enabled some remarkable achievements in hard control problems: with deep RL, agents have learned to <a href="https://arxiv.org/abs/1602.01783">play video games directly from pixels</a>, to control robots <a href="https://arxiv.org/abs/1502.05477">in simulation</a> and <a href="http://rll.berkeley.edu/deeplearningrobotics/">in the real world</a>, <a href="https://blog.openai.com/robots-that-learn/">to learn object manipulation from demonstrations</a>, and even to <a href="https://deepmind.com/research/publications/mastering-game-go-deep-neural-networks-tree-search/">beat human grandmasters at Go</a>. Hopefully, we’ll soon be able to take deep RL out of the lab and put it into practical, everyday technologies, like UAV control and household robots. But before we can do that, we have to address the most important concern: safety.</p>

<p>We recently developed a principled way to incorporate safety requirements and other constraints directly into a family of state-of-the-art deep RL algorithms. Our approach, Constrained Policy Optimization (CPO), makes sure that the agent satisfies constraints at every step of the learning process. Specifically, we try to satisfy constraints on <em>costs</em>: the designer assigns a cost and a limit for each outcome that the agent should avoid, and the agent learns to keep all of its costs below their limits.</p>

<p>This kind of <a href="http://www-sop.inria.fr/members/Eitan.Altman/TEMP/h.pdf">constrained RL</a> approach has been around for a long time, and has even inspired closely-related work here at Berkeley on <a href="https://arxiv.org/abs/1705.05394">probabilistically safe policy transfer</a>. But CPO is the first algorithm that makes it practical to apply deep RL to the constrained setting for general situations—and furthermore, it comes with theoretical performance guarantees.</p>

<p><a href="https://arxiv.org/abs/1705.10528">In our paper</a>, we describe an efficient way to run CPO, and we show that CPO can successfully train neural network agents to maximize reward while satisfying constraints in tasks with realistic robot simulations. <strong>If you want to try applying CPO to your constrained RL problem, <a href="https://github.com/jachiam/cpo">we’ve open-sourced our code.</a></strong></p>

<h1 id="why-do-we-need-constraints-for-safety">Why Do We Need Constraints for Safety?</h1>

<p>RL agents are trained to maximize a reward signal, which must be specified in advance by a human designer. If the reward signal isn’t properly designed, the agent <a href="https://blog.openai.com/faulty-reward-functions/">can learn unintended or even harmful behavior</a>. If it were easy to design reward functions, this wouldn’t be an issue, but unfortunately it’s fundamentally challenging; this is a key motivation for using constraints.</p>

<p>To illustrate, let’s consider a simplified example based on a real-world use case: a mobile robot is supposed to do some task (say, running in a circle), while staying inside of a safe area. We’ll consider the robot “safe” if its frequency of leaving the safe area is less than some pre-selected threshold, and otherwise, it’s “unsafe.”</p>

<p>This kind of problem is easy to describe in the constrained RL setting, in terms of both rewards (for running as fast as possible) and constraints (on frequency of leaving the safe area). But for standard RL, we run into difficulty because we have to design all behavior through the reward function alone. Mistakes in reward design could result in agents that are either too risk-averse,</p>

<p>
<img src="http://bair.berkeley.edu/static/blog/cpo/cpo1.gif">
</p>

<p>and therefore useless, or too risk-prone,</p>

<p>
<img src="http://bair.berkeley.edu/static/blog/cpo/cpo2.gif">
</p>

<p>and therefore dangerous. (Here, the “safe” area is between the blue panels.)</p>

<p>There’s another angle on why the standard RL approach is bad for safety. RL agents learn by trial and error, and they <em>explore</em> by trying many different policies before converging. So even if we design a reward function that leads an agent to safe policies at optimum, it could still result in unsafe exploration behavior somewhere between the beginning and end of training. For robots that have to learn in the real world, this is a serious problem.</p>

<p>Like in the example, autonomous systems in the real world are usually considered safe if failure modes are rare and happen less often than some pre-selected frequency. This motivates the constrained RL formulation as the natural way to incorporate safety into RL. Furthermore, we want to make sure that <em>every exploration policy</em> is also constraint-satisfying.</p>

<p>CPO is designed to meet these needs. Here’s what it came up with for the example task:</p>

<p>
<img src="http://bair.berkeley.edu/static/blog/cpo/cpo3.gif">
</p>

<h1 id="cpo-local-policy-search-for-constrained-rl">CPO: Local Policy Search for Constrained RL</h1>

<p>A standard way to learn policies is by local policy search, where we iteratively improve policies until convergence to optimum. This kind of search is called ‘local’ because each new policy is required to be close, in some sense, to the old one. To give an example, policy gradient methods are local policy search algorithms that keep policies close by only taking small steps in the direction of the gradient of performance. (If you’re unfamiliar with policy gradients, <a href="http://karpathy.github.io/2016/05/31/rl/">Andrej Karpathy</a> has a nice introduction!)</p>

<p><a href="https://arxiv.org/abs/1502.05477">Trust region methods</a> are another kind of local policy search algorithm. They also use policy gradients, but they make a special requirement for how policies are updated: each new policy has to be close to the old one in terms of <strong>average <a href="https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence">KL-divergence</a></strong>. KL-divergence is a measure of how different two probability distributions are from each other, and because policies output probability distributions over actions, KL-divergence is a natural way to measure “distance” between policies.</p>

<p>Trust region methods wind up performing extremely well in practice for training neural network policies, and are able to avoid the issue of performance collapse that standard policy gradient algorithms can sometimes suffer from (see <a href="https://arxiv.org/abs/1604.06778">Duan et al., 2016</a>).</p>

<p>CPO is a trust region method for constrained RL which approximately enforces the constraints in every policy update. It uses approximations of the constraints to predict how much the constraint costs might change after any given update, and then chooses the update that will most improve performance while keeping the constraint costs below their limits.</p>

<p>
<img src="http://bair.berkeley.edu/static/blog/cpo/cpo4.gif"><br>
<i>In this graphic, we illustrate what goes on under the hood of our CPO implementation. We start with the current policy iterate, the star at the center. The policy gradient (dark blue arrow) points in the direction that increases reward. The optimal step before considering safety constraints (dotted blue arrow) lies on the edge of the KL trust region (blue oval). But we only want to step in the constraint-satisfying area (light green half-space). So we adjust our step (to the dark green arrow), so that we're still increasing reward as much as possible but also staying safe. Our new policy after the CPO update is the star at the end.</i>
</p>

<p>We also have some nice theoretical results to complement the practical algorithm: we derive a new bound to describe the quality of our approximations, in terms of the average KL-divergence. This lets us prove a guarantee on the performance of trust region methods in general, which explains how well they do, and also lets us guarantee the worst-case constraint violation which is possible after a CPO update.</p>

<p>As for empirical performance: we find that CPO works quite well! In both the circle task (discussed above) and a more complex gathering task (where the agent wants to collect green apples, and is constrained to avoid red bombs), CPO learns good policies and approximately satisfies constraints all throughout training.</p>

<p>
<img src="http://bair.berkeley.edu/static/blog/cpo/cpo5.gif"><br>
<i>CPO learned a policy that collects as many green apples as possible, while not being allowed to collect more than 0.1 red bombs per trajectory on average.</i>
</p>

<h1 id="whats-next">What’s Next</h1>

<p>Our hope is that CPO will be a useful step towards bringing RL out of the lab and into the real world in a safe, reliable way. While we’re pushing forward on new research, there are a number of interesting ideas for how CPO might be applied which we invite the broader research community to consider and expand on. Things like</p>

<ul>
  <li>using CPO for safe transfer learning, to help move a policy learned in simulation into the real world,</li>
  <li>using CPO to learn safe policies from unsafe demonstrations,</li>
  <li>and improving the sample efficiency of CPO to minimize the number of errors an agent needs to make in order to learn from them.</li>
</ul>

<p>As a reminder, <strong>if you’re interested in pursuing any of these ideas or others, <a href="https://github.com/jachiam/cpo">feel free to use our code as a jumping-off point!</a></strong></p>

</article>

</div><br>
<br>
<br>
<br>
<br>
<br>
via The Berkeley Artificial Intelligence Research Blog https://ift.tt/2EPVVfk<br>
<br>
October 24, 2018 at 02:22PM
